mapie/calibration.py

```diff
@@ -0,0 +1,502 @@
+
+        - Top-Label method means that you take the maximum probability
+        and calibrated each class when it is the maximum separately.
+
+        Parameters
+        ----------
+        method : str
+            The method that will be taken into account.
+        pred : NDArray of shape (n_samples, n_classes)
+            The output from the predict_proba of the classifier.
+
+        Returns
+        -------
+        Tuple[NDArray, NDArray] of shapes (n_samples,) and (n_samples,)
+            The first element corresponds to the values that have to be
+            calibrated and the second element corresponds to class to be
+            associated.
+
+            In the Top-Label setting, the latter refers to the different
+            classes that have to be individually calibrated.
+
+        Raises
+        ------
+        ValueError
+            If the method provided has not been implemented.
+        """
+        if method == "top_label":
+            max_class_prob = np.max(pred, axis=1).reshape(-1, 1)
+            max_class_prob_arg = self.classes_[np.argmax(pred, axis=1)]
+        else:
+            raise ValueError("No other methods have been implemented yet.")
+        check_consistent_length(max_class_prob, max_class_prob_arg)
+        return max_class_prob, max_class_prob_arg
+
+    def _fit_calibrators(
+        self,
+        X: ArrayLike,
+        y: ArrayLike,
+        sample_weight: Optional[ArrayLike],
+        estimator: Union[ClassifierMixin, Pipeline],
+        calibrator: RegressorMixin,
+    ) -> Dict[int, RegressorMixin]:
+        """_summary_
+
+        Parameters
+        ----------
+        X : ArrayLike of shape (n_samples, n_features)
+            Training data.
+        y : ArrayLike of shape (n_samples,)
+            Training labels.
+        sample_weight : Optional[ArrayLike] of shape (n_samples,)
+            Sample weights for fitting the out-of-fold models.
+            If None, then samples are equally weighted.
+            If some weights are null,
+            their corresponding observations are removed
+            before the fitting process and hence have no residuals.
+            If weights are non-uniform, residuals are still uniformly weighted.
+            Note that the sample weight defined are only for the training, not
+            for the calibration procedure.
+            By default ``None``.
+        estimator : ClassifierMixin
+            Estimator fitted.
+        calibrator : RegressorMixin
+            Calibrator to train.
+
+        Returns
+        -------
+        Dict[int, RegressorMixin]
+            Dictionnary of fitted calibrators.
+        """
+        X, y = indexable(X, y)
+        y = _check_y(y)
+        y_pred_calib = estimator.predict_proba(X=X)
+        max_prob, max_prob_arg = self._get_labels(
+            cast(str, self.method),
+            y_pred_calib
+        )
+        calibrators = {}
+        for item in np.unique(max_prob_arg):
```

Comment:
you must iterate over estimator.classes_ : if a class is not predicted in your calib dataset, but it is in your test dataset it will result into an error. If it creates à empty dataset, return the identity calibrator (y_pred_proba = y_pred_proba_calib) and display a warning for the user. This special use case should be tested